### ISSUE TITLE
Will the training framework do upsampling when train-num-samples is far more than the amount of actual data

### ISSUE THREAD

**Issue by zwsjink** [mlfoundations/datacomp#55]
In the Datacomp paper, 
![image](https://github.com/mlfoundations/datacomp/assets/8237953/f7afa89b-7bd1-4d3a-adf0-f975342822f1)
there is only 14M data left  after _image-based filtering & clip score thresholding_   step under medium scale , however the train-num-samples  is equal to 128M/5epoch =  25600000 / epoch  , which is larger than  14M, so I suppose the open_clip  will do an upsampling to the data, am I correct?

**Comment by zwsjink** [comment]
by saying  "upsampling", I mean some data will appear more than once during an individual epoch.

**Comment by gabrielilharco** [comment]
Hey @zwsjink. Yes, that will happen (regardless of the sampling strategy, by the pigeonhole principle). Also note that the training script samples datapoints with replacement.

**Comment by zwsjink** [comment]
> Hey @zwsjink. Yes, that will happen (regardless of the sampling strategy, by the pigeonhole principle). Also note that the training script samples datapoints with replacement.

I see. So when you are saying   "with replacement" ,do you mean  this 
```
parser.add_argument(
        "--dataset-resampled",
        default=False,
        action="store_true",
        help="Whether to use sampling with replacement for webdataset shard selection."
    )
```
option,     so that   we not only shuffle between  shards but also within shards to achieve more randomness?

**Comment by gabrielilharco** [comment]
Yes, every time a worker needs a new shard, they sample uniformly from the pool of all shards. The samples in a shard are used to fill a shuffle buffer, from which samples are drawn at random.

**Comment by zwsjink** [comment]
> Yes, every time a worker needs a new shard, they sample uniformly from the pool of all shards. The samples in a shard are used to fill a shuffle buffer, from which samples are drawn at random.

 Thanks for the clarification. :D

**Comment by zwsjink** [comment]
> Yes, every time a worker needs a new shard, they sample uniformly from the pool of all shards. The samples in a shard are used to fill a shuffle buffer, from which samples are drawn at random.

sorry to bother, but what is the difference between two checkpoints i get after training ? I see one named epoch_5.pt and another epoch_latest.pt ,  I do see the original config for number of training epochs is equal to 5, so I believe these two checkpoints are same? but the file size looks different

**Comment by gabrielilharco** [comment]
That's a bit of a quirk from open_clip, the checkpoints should be the same

**Comment by gabrielilharco** [comment]
Closing for now, feel free to re-open if you have other questions!
